// auditjs/bin/Audit/AuditOSSIndex.js
"use strict";
/*
 * Copyright 2019-Present Sonatype Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditOSSIndex = void 0;
const Formatter_1 = require("./Formatters/Formatter");
const JsonFormatter_1 = require("./Formatters/JsonFormatter");
const TextFormatter_1 = require("./Formatters/TextFormatter");
const XmlFormatter_1 = require("./Formatters/XmlFormatter");
class AuditOSSIndex {
    constructor(quiet = false, json = false, xml = false) {
        this.quiet = quiet;
        this.json = json;
        this.xml = xml;
        if (json) {
            this.formatter = new JsonFormatter_1.JsonFormatter();
        }
        else if (xml) {
            this.formatter = new XmlFormatter_1.XmlFormatter();
        }
        else {
            this.formatter = new TextFormatter_1.TextFormatter(quiet);
        }
    }
    auditResults(results) {
        if (this.quiet) {
            results = results.filter((x) => {
                var _a;
                return x.vulnerabilities && ((_a = x.vulnerabilities) === null || _a === void 0 ? void 0 : _a.length) > 0;
            });
        }
        this.formatter.printAuditResults(results);
        return (0, Formatter_1.getNumberOfVulnerablePackagesFromResults)(results) > 0;
    }
}
exports.AuditOSSIndex = AuditOSSIndex;
