// auditjs/bin/Munchers/Bower.js
"use strict";
/*
 * Copyright 2019-Present Sonatype Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bower = void 0;
const Coordinates_1 = require("../Types/Coordinates");
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
class Bower {
    constructor(devDependencies = false) {
        this.devDependencies = devDependencies;
    }
    getSbomFromCommand() {
        throw new Error('Method not implemented.');
    }
    isValid() {
        const tempPath = path_1.default.join(process.cwd(), 'bower.json');
        return fs_1.default.existsSync(tempPath);
    }
    getDepList() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.getInstalledDeps();
        });
    }
    getInstalledDeps() {
        return __awaiter(this, void 0, void 0, function* () {
            const depsArray = [];
            const file = fs_1.default.readFileSync(path_1.default.join(process.cwd(), 'bower.json'));
            const json = JSON.parse(file.toString());
            Object.keys(json.dependencies).map((x) => {
                const version = json.dependencies[x];
                depsArray.push(new Coordinates_1.Coordinates(x, version.replace('~', ''), ''));
            });
            if (this.devDependencies) {
                Object.keys(json.devDependencies).map((x) => {
                    const version = json.devDependencies[x];
                    depsArray.push(new Coordinates_1.Coordinates(x, version.replace('~', ''), ''));
                });
            }
            return depsArray;
        });
    }
}
exports.Bower = Bower;
